jax/lax/lax_control_flow.py

```diff
@@ -1213,21 +1211,46 @@ def scan(f, init, xs, length=None):
     ys = tree_multimap(stack, *maybe_reversed(ys))
     return carry, ys
 
-  carry_avals = tuple(_map(_abstractify, init_flat))
   x_shapes = [masking.padded_shape_as_value(x.shape[1:]) for x in xs_flat]
   x_dtypes = [x.dtype for x in xs_flat]
   x_avals = tuple(_map(ShapedArray, x_shapes, x_dtypes))
-  jaxpr, consts, out_tree = _initial_style_jaxpr(f, in_tree, carry_avals + x_avals)
-  out_tree_children = out_tree.children()
-  if len(out_tree_children) != 2:
-    msg = "scan body output must be a pair, got {}."
-    raise TypeError(msg.format(tree_unflatten(out_tree, jaxpr.out_avals)))
-  # TODO(jakevdp): if any of carry_avals are weakly typed, promote them with output
-  # carry avals to get matching types & re-compile the jaxpr.
-  # Question to think about: will there be a fixed point?
+
+  # The carry input and output avals must match exactly. However, we want to account for
+  # the case when init contains weakly-typed values (e.g. Python scalars), with avals that
+  # may not match the output despite being compatible by virtue of their weak type.
+  # To do this, we compute the jaxpr in two passes: first with the raw inputs, and if
+  # necessary, a second time with modified init values.
+  for pass_ in ['first', 'second']:
+    init_flat, init_tree = tree_flatten(init)
+    in_flat, in_tree = tree_flatten((init, xs))
+
+    carry_avals = tuple(_map(_abstractify, init_flat))
+    jaxpr, consts, out_tree = _initial_style_jaxpr(f, in_tree, carry_avals + x_avals)
+    out_tree_children = out_tree.children()
+    if len(out_tree_children) != 2:
+      msg = "scan body output must be a pair, got {} in {} pass."
+      raise TypeError(msg.format(tree_unflatten(out_tree, jaxpr.out_avals), pass_))
+    carry_avals_out = jaxpr.out_avals[:out_tree_children[0].num_leaves]
+
+    # weak type logic run only on the first pass.
+    if pass_ != 'first':
+      break
```

Comment:
Another way to structure this code, rather than the loop over `['first', 'second']`, might be to factor the jaxpr formation part into a function, then just call it once, check for weak type mismatches, and call it again if needed. I think that might be cleaner than the loop with breaks. We can try that out later if you'd rather land this immediately, though.

Reply:
I initially did that, but the subsequent code uses 7 or 8 of the intermediate values computed in the function. Returning all of those seemed error-prone. I thought about defining a custom namedtuple to organize all of those, but that seems overly-engineered. I also thought about using `nonlocal` to let all those values leak passively into the parent scope, but that struck me as confusing. I finally landed on the loop approach as the least-worst option.

Reply:
I think a function returning a tuple is probably the best approach. I don't think it's error-prone because it's a function with effectively a single call site.

Reply:
OK, I'll make an update.

Reply:
#4512